src/commonMain/kotlin/io/github/koalaplot/core/bar/BarPlotShapes.kt

```diff
@@ -0,0 +1,327 @@
+                )
+            )
+        }
+        val cutoutArc = Path().apply {
+            addArc(
+                oval = Rect(
+                    offset = Offset(0F, shapeHeight - shapeWidth),
+                    size = Size(shapeWidth, shapeWidth)
+                ),
+                startAngleDegrees = 0F,
+                sweepAngleDegrees = 180F
+            )
+        }
+        val cutout = (cutoutRect - cutoutArc)
+        return (outline.path - cutout).let(Outline::Generic)
+    }
+}
+
+/**
+ * Rectangle shape with concave/convex shaped sides.
+ * Useful for Single Vertical Bar and Stacked Bars Plot rendering.
+ *
+ * @param xyGraphScope Provides access to [yAxisModel] and acts as an implementation of [XYGraphScope].
+ * @param value The [VerticalBarPlotEntry] that defines the cutouts for the [ConcaveConvexShape].
+ */
+@Stable
+public class ConcaveConvexShape<X, E : VerticalBarPlotEntry<X, Float>>(
+    private val xyGraphScope: XYGraphScope<X, Float>,
+    private val index: Int,
+    private val value: E
+) : Shape, XYGraphScope<X, Float> by xyGraphScope {
+    override fun createOutline(
+        size: Size,
+        layoutDirection: LayoutDirection,
+        density: Density
+    ): Outline {
+        val shapeWidth = size.width
+        val shapeHeight = size.height
+        val arcRadius = shapeWidth / 2
+
+        // Rendering negative values
+        val isInverted = value.y.yMax < value.y.yMin
+        // Required for proper bar rendering in waterfall charts
+        if (index == 0) {
+            val outline = PlaneConvexShape.createOutline(size, layoutDirection, density) as Outline.Generic
+
+            outline.path.apply {
+                // Rendering bar in negative direction
+                if (isInverted) {
+                    inverted(pivotX = shapeWidth / 2F, pivotY = shapeHeight / 2F)
+                }
+            }
+            return outline
+        }
+
+        val (yZeroOffset, yMinOffset, yMaxOffset) = yAxisModel.yOffsets(value.y.yMin, value.y.yMax)
+
+        // Prevent division by zero
+        return if (yMaxOffset == yMinOffset) {
+            Path().let(Outline::Generic)
+        } else {
+            val heightOffsetRatio = size.height / (yMaxOffset - yMinOffset)
+            val offsetToHeight = { offset: Float -> offset * heightOffsetRatio }
+
+            val yMinZeroOffset = yMinOffset - yZeroOffset
+            val yMaxZeroOffset = yMaxOffset - yZeroOffset
+
+            val yMinZeroHeight = offsetToHeight(yMinZeroOffset)
+            val yMaxZeroHeight = offsetToHeight(yMaxZeroOffset)
+
+            val yMinZeroArcHeight = max((arcRadius - yMinZeroHeight), 0F)
+            val yMaxZeroArcHeight = max((arcRadius - yMaxZeroHeight), 0F)
+
+            val yMaxZeroArcHeightDegrees =
+                asin(yMaxZeroArcHeight / arcRadius).rad.toDegrees().value.toFloat()
+
+            Path().apply {
+                (
+                    Path().apply {
```

Comment:
Can you add comments describing the purpose of these calculations? The default bar doesn't need to know where it is within the graph. Is this needed to shape the bottom of a bar if there is a short bar under it?